Data/Vector/Primitive.hs

```diff
@@ -183,11 +186,22 @@ import Data.Traversable ( Traversable )
 #endif
 
 #if __GLASGOW_HASKELL__ >= 708
+import Data.Coerce
+import Unsafe.Coerce
 import qualified GHC.Exts as Exts
 #endif
 
 #if __GLASGOW_HASKELL__ >= 708
-type role Vector representational
+type role Vector nominal
+
+-- | /O(1)/ Unsafely coerce a mutable vector from one element type to another,
+-- representationally equal type. The operation just changes the type of the
+-- underlying pointer and does not modify the elements.
+--
+-- While it guarantees that elements of vector are representationally
+-- equal it doesn't guarantee that 'Prim' instances are memory-compatible.
```

Comment:
This is a subtle subject, and an unprepared reader may be well confused how types could happen to be "representationally equal", but not "memory-compatible". Here is my take (still far from perfect):

"Coerce a vector to another vector, when types of underlying elements are 'Coercible' to each other (e. g., newtypes). This is unsafe, because 'Prim' instances of newtypes can potentially still have different memory representations."